sign/mayo/mode1/internal/matrix.go

```diff
@@ -0,0 +1,289 @@
+				vecAddPacked(P, pst[P*(c*K+k):], accumulator[r*K+k][P*int(s[r*N+c]):])
+			}
+		}
+	}
+
+	for i := 0; i < K*K; i++ {
+		aggregate(P, accumulator[i], sps[P*i:])
+	}
+}
+
+// p is always P, but is still kept to be consistent with other functions
+//
+//nolint:unparam
+func vecAddPacked(p int, in []uint64, acc []uint64) {
+	for i := 0; i < p; i++ {
+		acc[i] ^= in[i]
+	}
+}
+
+func aggregate(p int, bins [P * 16]uint64, out []uint64) {
+	// The following two methods are mathematically equivalent, but the second one is slightly faster.
+
+	// the powers of x mod x^4+x+1, represented as integers, are 1,2,4,8,3,..,13,9
+	// out = bins[9]*x^14 + bins[13]*x^13 + bins[15]*x^12 ... + bin[4]*x^2 + bins[2]*x + bins[1]
+	//     = ((bins[9]x+bins[13])x+bins[15])x + ... bins[4])x+bins[2])x+bins[1]
+
+	// vecMulAddPackedByX(p, bins[P*9:], bins[P*13:])
+	// vecMulAddPackedByX(p, bins[P*13:], bins[P*15:])
+	// vecMulAddPackedByX(p, bins[P*15:], bins[P*14:])
+	// vecMulAddPackedByX(p, bins[P*14:], bins[P*7:])
+	// vecMulAddPackedByX(p, bins[P*7:], bins[P*10:])
+	// vecMulAddPackedByX(p, bins[P*10:], bins[P*5:])
+	// vecMulAddPackedByX(p, bins[P*5:], bins[P*11:])
+	// vecMulAddPackedByX(p, bins[P*11:], bins[P*12:])
+	// vecMulAddPackedByX(p, bins[P*12:], bins[P*6:])
+	// vecMulAddPackedByX(p, bins[P*6:], bins[P*3:])
+	// vecMulAddPackedByX(p, bins[P*3:], bins[P*8:])
+	// vecMulAddPackedByX(p, bins[P*8:], bins[P*4:])
+	// vecMulAddPackedByX(p, bins[P*4:], bins[P*2:])
+	// vecMulAddPackedByX(p, bins[P*2:], bins[P*1:])
+	// copy(out[:P], bins[P*1:])
+
+	// In the reversed order of the above, because /x turns out to be slightly faster than *x.
+	// out = ((bins[2]x^-1+bins[4])x^-1+bins[8])x^-1 + ... bins[13])x^-1+bins[9])x^-1+bins[1]
+	vecMulAddPackedByInvX(p, bins[P*2:], bins[P*4:])
+	vecMulAddPackedByInvX(p, bins[P*4:], bins[P*8:])
+	vecMulAddPackedByInvX(p, bins[P*8:], bins[P*3:])
+	vecMulAddPackedByInvX(p, bins[P*3:], bins[P*6:])
+	vecMulAddPackedByInvX(p, bins[P*6:], bins[P*12:])
+	vecMulAddPackedByInvX(p, bins[P*12:], bins[P*11:])
+	vecMulAddPackedByInvX(p, bins[P*11:], bins[P*5:])
+	vecMulAddPackedByInvX(p, bins[P*5:], bins[P*10:])
+	vecMulAddPackedByInvX(p, bins[P*10:], bins[P*7:])
+	vecMulAddPackedByInvX(p, bins[P*7:], bins[P*14:])
+	vecMulAddPackedByInvX(p, bins[P*14:], bins[P*15:])
+	vecMulAddPackedByInvX(p, bins[P*15:], bins[P*13:])
+	vecMulAddPackedByInvX(p, bins[P*13:], bins[P*9:])
+	vecMulAddPackedByInvX(p, bins[P*9:], bins[P*1:])
+	copy(out[:P], bins[P*1:])
+}
+
+// func vecMulAddPackedByX(p int, in []uint64, acc []uint64) {
+// 	// vecMulAddPacked(p, in, 2, acc)
+
+// 	msb := uint64(0x8888888888888888)
+// 	for i := 0; i < p; i++ {
+// 		t := in[i] & msb
+// 		acc[i] ^= ((in[i] ^ t) << 1) ^ ((t >> 3) * 3)
+// 	}
+// }
+
+func vecMulAddPackedByInvX(p int, in []uint64, acc []uint64) {
+	// vecMulAddPacked(p, in, 9, acc)
+
+	lsb := uint64(0x1111111111111111)
+	for i := 0; i < p; i++ {
+		t := in[i] & lsb
+		acc[i] ^= ((in[i] ^ t) >> 1) ^ (t * 9)
+	}
+}
```

Comment:
This is Figure 2 method 3 right? Wouldn't method 2 be faster as we're not using AVX2 instructions?

Reply:
Here we chose to multiply by x^-1 all the way through, unlike Method 3 which interleaves *x and *x^-1 which probably gives more parallelism on more complex CPUs. On my MacBook M1 Pro, Method 2 was not faster.